chain/network/src/routing.rs

```diff
@@ -855,6 +857,176 @@ impl Graph {
+        let id0 = self.get_id(&peer0);
+        let id1 = self.get_id(&peer1);
+
+        self.adjacency.get_mut(&id0).unwrap().remove(&id1);
+    }
+
+    pub fn add_edge(&mut self, peer0: PeerId, peer1: PeerId) {
+        if !self.contains_edge(&peer0, &peer1) {
+            self.add_directed_edge(peer0.clone(), peer1.clone());
+            self.add_directed_edge(peer1, peer0);
+            self.total_active_edges += 1;
+        }
+    }
+
+    pub fn remove_edge(&mut self, peer0: &PeerId, peer1: &PeerId) {
+        if self.contains_edge(&peer0, &peer1) {
+            self.remove_directed_edge(&peer0, &peer1);
+            self.remove_directed_edge(&peer1, &peer0);
+            self.total_active_edges -= 1;
+        }
+    }
+
+    // TODO(MarX, #1363): This is too slow right now. (See benchmarks)
+    /// Compute for every node `u` on the graph (other than `source`) which are the neighbors of
+    /// `sources` which belong to the shortest path from `source` to `u`. Nodes that are
+    /// not connected to `source` will not appear in the result.
+    pub fn calculate_distance(&self) -> HashMap<PeerId, HashSet<PeerId>> {
+        let mut queue = vec![];
+        let mut distance: HashMap<u64, i32> = HashMap::new();
+        // TODO(MarX, #1363): Represent routes more efficiently at least while calculating distances
+        //let mut routes: HashMap<u32, HashSet<u32>> = HashMap::new();
+        let mut routes: HashMap<u64, u128> = HashMap::new();
+
+        // let source = self.source.clone();
+        let source_id = 0; //self.get_id(&source);
+        distance.insert(source_id, 0);
+
+        // Add active connections
+
+        if let Some(neighbors) = self.adjacency.get(&source_id) {
+            let mut id = 0;
+            for neighbor in neighbors {
+                if id >= 128 {
+                    break;
+                }
+
+                let mask = (1 as u128) << id;
+
+                queue.push(*neighbor);
+                distance.insert(*neighbor, 1);
+                routes.insert(*neighbor, mask); //vec![neighbor.clone()].drain(..).collect());
+
+                id += 1;
+            }
+        }
+
+        let mut head = 0;
+
+        while head < queue.len() {
+            let cur_peer = queue[head];
+            let cur_distance = *distance.get(&cur_peer).unwrap();
+            head += 1;
+
+            if let Some(neighbors) = self.adjacency.get(&cur_peer) {
+                for neighbor in neighbors {
+                    if let Entry::Vacant(entry) = distance.entry(*neighbor) {
+                        queue.push(*entry.key());
+                        entry.insert(cur_distance + 1);
+                        routes.insert(neighbor.clone(), 0);
+                    }
+                    // If this edge belong to a shortest path, all paths to
+                    // the closer nodes are also valid for the current node.
+                    if *distance.get(neighbor).unwrap() == cur_distance + 1 {
```

Comment:
Seems like this can be written as a single `*entry().get_or_insert()` to avoid double lookup in the hashmap and unwrap.

Reply:
Yes, or we could make distance/routes into an array once I make other changes